cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java

```diff
@@ -0,0 +1,575 @@
+    assertEquals(
+        Objects.requireNonNull(sourceCredentials.getAccessToken()).getExpirationTime(),
+        intermediateAccessToken.getExpirationTime());
+  }
+
+  @Test
+  public void refreshCredentialsIfRequired_firstCallWillFetchIntermediateCredentials()
+      throws IOException {
+    GoogleCredentials sourceCredentials =
+        getServiceAccountSourceCredentials(mockTokenServerTransportFactory);
+
+    ClientSideCredentialAccessBoundaryFactory factory =
+        ClientSideCredentialAccessBoundaryFactory.newBuilder()
+            .setSourceCredential(sourceCredentials)
+            .setHttpTransportFactory(mockStsTransportFactory)
+            .build();
+
+    // Verify that the first call to refreshCredentialsIfRequired() triggers a fetch of intermediate
+    // credentials, resulting in one request to the STS endpoint. This happens because the
+    // intermediate credentials are initially null.
+    assertEquals(0, mockStsTransportFactory.transport.getRequestCount());
+    factory.refreshCredentialsIfRequired();
+    assertEquals(1, mockStsTransportFactory.transport.getRequestCount());
+  }
+
+  @Test
+  public void refreshCredentialsIfRequired_noRefreshNeeded() throws IOException {
+    final ClientSideCredentialAccessBoundaryFactory factory =
+        getClientSideCredentialAccessBoundaryFactory(RefreshType.NONE);
+
+    // Call refreshCredentialsIfRequired() once to initialize the intermediate credentials. This
+    // should make one request to the STS endpoint.
+    factory.refreshCredentialsIfRequired();
+
+    // Verify that a subsequent call to refreshCredentialsIfRequired() does NOT trigger another
+    // refresh, as the token is still valid. The request count should remain the same.
+    assertEquals(1, mockStsTransportFactory.transport.getRequestCount());
+    factory.refreshCredentialsIfRequired();
+    assertEquals(1, mockStsTransportFactory.transport.getRequestCount());
+  }
+
+  @Test
+  public void refreshCredentialsIfRequired_blockingSingleThread() throws IOException {
+    final ClientSideCredentialAccessBoundaryFactory factory =
+        getClientSideCredentialAccessBoundaryFactory(RefreshType.BLOCKING);
+
+    // Call refreshCredentialsIfRequired() once to initialize the intermediate credentials. This
+    // should make one request to the STS endpoint.
+    factory.refreshCredentialsIfRequired();
+    assertEquals(1, mockStsTransportFactory.transport.getRequestCount());
+
+    // Simulate multiple calls to refreshCredentialsIfRequired. In blocking mode, each call should
+    // trigger a new request to the STS endpoint.
+    int numRefresh = 3;
+    for (int i = 0; i < numRefresh; i++) {
+      factory.refreshCredentialsIfRequired();
+    }
+
+    // Verify that the total number of requests to the STS endpoint is the initial request plus the
+    // number of subsequent refresh calls.
+    assertEquals(1 + numRefresh, mockStsTransportFactory.transport.getRequestCount());
```

Comment:
This is only the case for the test because we specify that all subsequent calls will be BLOCKING? IIUC, normally the first request would be blocking and the other three calls would be NONE since the token should be valid?

Reply:
Yes, exactly.